hadrons/common/continous_beam.py

```diff
@@ -0,0 +1,131 @@
+            )
+
+            self.number_of_tracks = max(1, self.number_of_tracks)
+
+            self.track_distribution = create_track_distribution(
+                self.computation_time_steps,
+                self.dt,
+                self.number_of_tracks,
+                self.separation_time_steps,
+                self.random_generator,
+            )
+
+        def get_number_of_tracks(self, time_step: int) -> bool:
+            if time_step == 0:
+                print(self.track_distribution)
+                return 1
+            return self.track_distribution[time_step]
+
+        def get_random_xy_coordinate(self):
+            return self.random_generator.random() * self.no_xy
+
+        def get_random_coordinates(self):
+            x = self.get_random_xy_coordinate()
+            y = self.get_random_xy_coordinate()
+
+            # check whether the point is inside the circle.
+            # it is too time comsuming simply to set x=rand(0,1)*no_xy
+            while (
+                sqrt((x - self.xy_middle_idx) ** 2 + (y - self.xy_middle_idx) ** 2)
+                > self.inner_radius
+            ):
+                x = self.get_random_xy_coordinate()
+                y = self.get_random_xy_coordinate()
+
+            return x, y
+
+        def get_track_for_time_step(self, time_step: int):
+            positive_array = np.zeros((self.no_xy, self.no_xy, self.no_z_with_buffer))
+            negative_array = np.zeros((self.no_xy, self.no_xy, self.no_z_with_buffer))
+            no_initialised_charge_carriers = 0.0
+
+            x, y = self.get_random_coordinates()
+
+            for k in tqdm(
+                range(self.no_z_electrode, self.no_z + self.no_z_electrode),
+                desc="Calculating track...",
+            ):
+                for i in range(self.no_xy):
+                    x_track_dist_squared = (i - x) ** 2
+                    x_chamber_dist_squared = (i - self.xy_middle_idx) ** 2
+
+                    for j in range(self.no_xy):
+                        distance_from_center = (
+                            sqrt(x_track_dist_squared + (j - y) ** 2)
+                            * self.grid_spacing
+                        )
+                        ion_density = self.Gaussian_factor * exp(
+                            -(distance_from_center**2) / self.track_radius**2
+                        )
+                        positive_array[i, j, k] += ion_density
+                        negative_array[i, j, k] += ion_density
+                        # calculate the recombination only for charge carriers inside the circle
+                        if (
+                            sqrt(x_chamber_dist_squared + (j - self.xy_middle_idx) ** 2)
+                            < self.inner_radius
+                        ):
+                            # TODO: I'm not sure why this check is here - when separatioon_time_steps >= computation_time_steps,
```

Comment:
@jbrage I've copied this `if` from the [cython implementation](https://github.com/jbrage/IonTracks/blob/master/hadrons/cython_files/continuous_beam.pyx#L211), but I'm not sure why it's here. It currently discards the initialized particles from the first few iterations - if any beams create charged particles during that time, they won't be accounted for in the result. Is this correct?

Reply:
@retkiewi yes, this actually is correct. The time it takes charged particles to drift half the electrode gap distance is an important characteristic time parameter here, because it is related to the steady-state case: Before that "separation time", there is a charge build-up giong on in the chamber. After that time, the charge in the ionization chamber is approximately constant, and we are generally only interested in the recombination at this steady-state level, i.e. the fraction of recombination after the "separation_time_steps".

So essentially, we should do the bookkeeping after "separation_time_steps", where we would count (i) the number of charge initialised, (ii), the charge collected, and calculate/count (iii) the fraction of charge recombination, i.e. the recombination factor

does that make sense?

Reply:
Yeah, that makes sense. Thanks! 😄